### ISSUE TITLE
Best way implement a waitForUrl method?

### ISSUE THREAD

**Issue by ankurkaushal** [webdriverio/webdriverio#575]
As my question states, I want to implement a wait for url method which basically wait for a url to load after clicking a button? (Imagine a scenario where you have created a record & the system takes you back to the dashboard).

**Comment by christian-bromann** [comment]
I will just copy past my answer from Gitter :wink: 
with v2.x it’s not straightforward do do that but in v3you can do the following:

```js
client.waitUntil(function() {
    return cleint.url().then(function(url) {
        return url === ‘http://mywebsite.com';
    });
}, 5000);
```

If you need something right now check out the [`waitForExist`](https://github.com/webdriverio/webdriverio/blob/master/lib/commands/waitForExist.js) command. You can get it working to wait on a certain url with some minor tweaks

**Comment by monolithed** [comment]
+1. It would be nice to do something like

waitForUrl(url).should.be.eventually.equal('....')

**Comment by ResolverJay** [comment]
It would be nice to have this built in, though it is not hard to create a custom command like this. (This is my very first custom function :) )
```
browser.addCommand('waitForUrl', function (url) {
    return this.waitUntil(function () {
        return this.url(function (err, res) {
            return res.value === url;
        })
    })
})
```

**Comment by monolithed** [comment]
@ResolverJay, 

```js
browser
    .url('http://facebok.com')
    .waitUntil('http://google.com').should.be.eventually.true
```

What result do you expect? 
`'http://facebok.com' != 'http://google.com'` ? — No! 
It will be something like:

`Promise was rejected with the following reason: timeout`
or
`Promise never resolved with an truthy value`

**Comment by ResolverJay** [comment]
@monolithed not sure if I understand your question. Do you want to compare the url at the end of the above custom function and return it? If this is the case, you can simply add it at the end of the function like this.
```
client.addCommand('waitForUrl', function (url) {
    return this.waitUntil(function () {
        return this.url(function (err, res) {
            return res.value === url;
        })
    }).url(function(err, res) {
        return res.value === url;
    })
})
```

Though, I personally don't like the idea of comparing the url in this function.
1. It already waits for the url, so it will have the right url (otherwise it will fail). Why bother to confirm the url? (url redirect?)
2. The original function returns browser, so you can continue chaining actions.
3. Keep each function simple on one purpose. If you want to compare the url, you can simply create another custom function like below that simply returns the url, so you can use it anywhere you want.
```
browser.addCommand('hasUrl', function (url) {
    return this.url(function(err, res) {
        return res.value;
    })
})
```

**Comment by monolithed** [comment]
> @ResolverJay,  Do you want to compare the url at the end of the above custom function and return it?
I mean, what error message do you expect to receive? 

> If you want to compare the url, you can simply create another custom function like below that simply returns the url, so you can use it anywhere you want.

Sounds good.

**Comment by ResolverJay** [comment]
@monolithed If I ran
```
browser
    .url('http://facebok.com')
    .waitForUrl('http://google.com')
```
I received
```
Promise never resolved with an truthy value
```
which I'm fine with.

**Comment by monolithed** [comment]
@ResolverJay, but no for me. 

```
http://example.com (start) -> 
           http://accounts.google.com (submit a request for an authorization)
<- http://example.com/some_page (end)
```

I want to see a place (url) where the error is thrown:
(http://example.com,  http://accounts.google.com or http://example.com/some_page).

**Comment by ResolverJay** [comment]
@monolithed I agree, it would be nice to see the URLs in the message. Since this is a custom function, you can inject your own error message. But if you want WebdriverIO to give a specific error message with details inside a custom function, I don't think that's feasible.

BTW, I'm not the right person to ask if you want a specific feature. I'm just sharing my knowledge/experience on how to use it here.

**Comment by OliverJAsh** [comment]
I would like to bump the idea of adding this as a command.

It might be useful to somebody, here is my version:

``` js
import compose from 'lodash/flowRight';

export const waitForURL = (expectedUrl) => {
  browser.waitUntil(
    () => browser.getUrl() === expectedUrl,
    // Use default timeout
    null,
    `Expected URL to be ${expectedUrl} but got ${browser.getUrl()}`
  );
};
export const waitForPath = compose(waitForURL, getUrl);
```

**Comment by monolithed** [comment]
@OliverJAsh, here is my version:


```js
/**
 * @alias browser.waitForUrl
 * @param {string|RegExp|Function} value
 * @param {number} timeout — ms
 * @param {number} revert
 * @returns {boolean}
 */
browser.addCommand('waitForUrl', function (value, timeout, revert) {
	let url, actual;

	try {
		return browser.waitUntil(() => {
			url = browser.getUrl();
			actual = value === url;

			// This slash is added by Selenium
			if (typeof value === 'string' && !value.endsWith('/')) {
				url = url.replace(/\/$/, '');
			}

			if (typeof value === 'function') {
				actual = value(url);
			} else if (value[Symbol.match]) {
				actual = value.test(url);
			}

			if (revert) {
				actual = !actual;
			}

			return value && actual;
		}, timeout, '');
	} catch (error) {
		let message = 'Could not wait for required url:';
			message += `\n\tActual: ${url}`;
			message += `\n\tExpected: ${value}`;

		throw new Error(message);
	}
});
```
